README
WHAT IS LIBOOBS?
================

Author: Carlos Garnacho  <carlosg@gnome.org>

Liboobs is a wrapping library to the System Tools Backends, it will provide easy to access GObjects to system configuration details, like users, groups and network interfaces, it will handle sessions with the backend and data consistency too

To compile, you will need system-tools-backends 2.0 and DBus >= 0.31

WHERE DOES THIS WEIRD NAME COME FROM?
=====================================

Blame Evandro's (aka hk) and Alberto's (aka arc) blackened minds :)



NEWS
Liboobs 3.0.0, 2010-04-02
-----------------------------------------------
Liboobs version 3.0.0 "Void" has been released.

  - Update DOAP file.


Liboobs 2.91.1, 2010-12-19
-----------------------------------------------
Liboobs version 2.91.1 "Strict Groups" has been released.

  - Fix updating main groups for new users
  - Remove idle tasks when destroying objects
  - Link gtk-doc scanner against OOBS_LIBS so strict linking works


Liboobs 2.32.0, 2010-09-27
-----------------------------------------------
Liboobs version 2.32.0 "If It Ain't Broke, Don't Fix It" has been released.

No changes release to use the stable version number.


Liboobs 2.31.91, 2010-08-30
-----------------------------------------------
Liboobs version 2.31.91 "Better Break Late Than Never" has been released.

  - Add new values to OobsUserHomeFlags to allow more operations
    on home directory
  - Update documentation for OOBS_USER_ERASE_HOME, whose signification was extended
  - Add oobs_user_get_full_name_fallback(), providing always meaningful names
  - Set D-Bus timeout to max for async replies
  - Bump soname to 5.0.0 to reflect ABI break of last cycle (too late)

Liboobs 2.31.1, 2010-05-27
-----------------------------------------------
Liboobs version 2.31.1 "Private" has been released.

  - Clear runlevels list before updating (fixes enabling NTP the first
time in time-admin)
  - Don't include dbus/dbus.h in public headers (fixes building clients)
  - Add libdubus-1 to Requires.private


Liboobs 2.30.0, 2010-03-29
-----------------------------------------------
Liboobs version 2.30.0 "Don't You Notice It's Freaking Good?" has been released.

  - Fix removing deleted users from group members (typo)
  - Don't forget to commit minimum and maximum UIDs/GIDs
  - Use guint32 instead of uid_t with D-Bus to avoid potential size problems


Liboobs 2.29.92, 2010-03-08
-----------------------------------------------
Liboobs version 2.29.92 "Midnight" has been released.

   - Fix 'password-disabled' property (account lock support)


Liboobs 2.29.91, 2010-02-15
-----------------------------------------------
Liboobs version 2.29.91 "Noon" has been released.

- Various crash fixes
- Improve references handling around OobsObjects
- Fix appending user home flags to DBusMessage


Liboobs 2.29.90, 2010-02-08
-----------------------------------------------
Liboobs version 2.29.90 "Frosty" has been released.

- OobsService is now an OobsObject, which should normally be committed separately
- Don't remove unknown users from /etc/group (LDAP compliance)
- Disable HAL support by default (Javier Jardón)
- Make all OobsObjects real singletons
- Ensure passwords are erased from memory
- Fix reference counting with OobsUsers and OobsGroups


Liboobs 2.29.3, 2010-01-25
-----------------------------------------------
Liboobs version 2.29.3 "Tune Up" has been released.

- New function oobs_object_authenticate() allowing clients to request PolicyKit authentication to be more user-friendly
- Basic GError reporting support, used in the above function for now
- Fix OobsSelfConfig getting the UID of current user, and add oobs_self_config_is_user_self()
- Move from utmp to utmpx (thanks to Ed Schouten)


Liboobs 2.29.2.1, 2010-01-12
-----------------------------------------------
Liboobs version 2.29.2.1 "I Got It Bad And That Ain't Good" has been released.

- Fix missing header that was preventing the GNOME System Tools from building (Götz Waschk)


Liboobs 2.29.2, 2010-01-11
-----------------------------------------------

Liboobs version 2.29.2 "Follow Suit" has been released. This release depends on the System Tools Backends 2.9.0.

- Move to the new System Tools Backends users and groups protocol. This mainly implies that OobsUser and OobsGroup are now OobsObjects, and can be committed individually, with new add() and delete() methods.
- Add new fields from this protocol: passwords are now in clear text, UIDs and GIDs can be guint32, new password and home flags, encrypted home boolean, locale and location.
- When committed, OobsUser is immediately updated with the D-Bus reply, to fill fields that were left empty and chosen by the platform tools.
- Add a lot of convenience functions to manage users and groups lists, taken from the GNOME System Tools.
- Raise OOBS_MAX_[UG]ID to G_MAXUINT32 on platforms where it is possible, now that the protocol supports it. Use << instead of pow().
- Fix OobsSession not being a singleton.

Liboobs 2.29.1, 2009-11-30
-----------------------------------------------

Liboobs version 2.29.1 "Not Yet!" has been released.

- Remember init scripts priority, allowing for GUIs not to care about them, simply passing 0 to mean "previous".
- Don't allow clients to remove init scripts by setting their status to "ignore": scripts should always use "start" or "stop" statuses, so that their priority is not lost when disabling them.
- Print a warning to the console when asynchronous commit fails. This ensures that errors will always be reported for debugging purposes.


Liboobs 2.22.2, 2009-09-06
-----------------------------------------------

Liboobs version 2.22.2 "Where's Waldo?" has been released.

Liboobs is a lightweight library that provides a GObject based interface to system-tools-backends. It's completely abstracted of the communication and authentication details, making it easy for applications to integrate with the system details.

Changes since last release
==========================

- Print error from D-Bus in case of failure on start. This should greatly help debugging.
- Don't consider empty user password as invalid.
- Make gtk-doc code documentation use inlined section comments, so that it is updated on every change. This will ensure the documentation does not get outdated again.

Thanks to anyone involved!

Downloading
===========
You can get it from :
ftp://ftp.gnome.org/pub/GNOME/sources/liboobs/2.22/


Liboobs 2.22.1, 2009-08-16
-----------------------------------------------


src/md5.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/* $Id$
 *
 * This code implements the MD5 message-digest algorithm.
 * The algorithm is due to Ron Rivest.  This code was
 * written by Colin Plumb in 1993, no copyright is claimed.
 * This code is in the public domain; do with it what you wish.
 *
 * Equivalent code is available from RSA Data Security, Inc.
 * This code has been tested against that, and is equivalent,
 * except that you don't need to include two pages of legalese
 * with every copy.
 *
 * To compute the message digest of a chunk of bytes, declare an
 * MD5Context structure, pass it to MD5Init, call MD5Update as
 * needed on buffers full of bytes, and then call MD5Final, which
 * will fill a supplied 16-byte array with the digest.
 *
 * $Log$
 * Revision 1.1  2005/08/13 16:35:38  carlosg
 * 2005-08-13  Carlos Garnacho Parro  <carlosg@gnome.org>
 *
 *         * src/md5-crypt.c, md5.[ch]: added to support MD5 crypt
 *         * src/oobs-defines.h: private header for adding internal defines
 *         * src/oobs-groupsconfig.[ch]: added. Object for handing groups
 *         configuration
 *         * src/oobs-list.[ch]: now it's an independent list, inheriting for
 *         GObject, it's done because a single OobsObject might contain several
 *         lists
 *         * src/oobs-list-private.h: added. Private header for holding
 *         OobsList constructor, this won't be made public, as API users
 *         shouldn't create/destroy it, just use what the OobsObjects provide
 *         * src/oobs-ntpconfig.[ch]: added. Object to represent NTP servers
 *         configuration
 *         * src/oobs-ntpserver.[ch]: added. Object to represent a NTP server
 *         to which the computer connects
 *         * src/oobs-shellsconfig.[ch]: added. Object to represent Shells
 *         configuration
 *         * src/oobs-usersconfig.[ch]: added. object to handle users
 *         configuration
 *         * src/oobs.h: one header to include them all
 *         * src/oobs-group.c: added "crypted-password" property, complete API
 *         * src/oobs-object.c, oobs-session.c, oobs-share-nfs.c,
 *         oobs-share-smb.c, oobs-share.c, oobs-shareslist.c, oobs-shell.c,
 *         oobs-shellslist.c,: code fixes and other improvements
 *         * src/oobs-shell.[ch]: complete API
 *         * src/oobs-user.[ch]: ditto
 *         * configure.in: add checks for libcrypt
 *         * src/Makefile.am: add all these files
 *
 * Revision 1.2  2001/03/02 01:31:24  unammx
 * Added emacs style command line at the beginning of all [hc] files, except for the archiver files, which already have their own.
 *
 * Revision 1.1  2000/11/08 03:44:39  unammx
 * md5 crypt code.
 *
 * Revision 1.1.1.3  2000/01/03 17:34:46  gafton
 * Imported Linux-PAM-0.68
 *
 * Revision 1.2  1999/07/04 23:22:38  morgan
 * Andrey's MD5 (bigendian) work around + cleanup to address problems with
 * applications that let an (ab)user kill them off without giving PAM the
 * opportunity to end. [Problem report from Tani Hosokawa on bugtraq.]
 *
 * Revision 1.1.1.1  1998/07/12 05:17:17  morgan
 * Linux PAM sources pre-0.66
 *
 * Revision 1.1  1996/09/05 06:43:31  morgan
 * Initial revision
 *
 */

#include <string.h>
#include "md5.h"

#ifndef HIGHFIRST
#define byteReverse(buf, len)	/* Nothing */
#else
static void byteReverse(unsigned char *buf, unsigned longs);

#ifndef ASM_MD5
/*
 * Note: this code is harmless on little-endian machines.
 */
static void byteReverse(unsigned char *buf, unsigned longs)
{
    uint32 t;
    do {
	t = (uint32) ((unsigned) buf[3] << 8 | buf[2]) << 16 |
	    ((unsigned) buf[1] << 8 | buf[0]);
	*(uint32 *) buf = t;
	buf += 4;
    } while (--longs);
}
#endif
#endif

/*
 * Start MD5 accumulation.  Set bit count to 0 and buffer to mysterious
 * initialization constants.
 */
void MD5Init(struct MD5Context *ctx)
{
    ctx->buf[0] = 0x67452301U;
    ctx->buf[1] = 0xefcdab89U;
    ctx->buf[2] = 0x98badcfeU;
    ctx->buf[3] = 0x10325476U;

    ctx->bits[0] = 0;
    ctx->bits[1] = 0;
}

/*
 * Update context to reflect the concatenation of another buffer full
 * of bytes.
 */
void MD5Update(struct MD5Context *ctx, unsigned const char *buf, unsigned len)
{
    uint32 t;

    /* Update bitcount */

    t = ctx->bits[0];
    if ((ctx->bits[0] = t + ((uint32) len << 3)) < t)
	ctx->bits[1]++;		/* Carry from low to high */
    ctx->bits[1] += len >> 29;

    t = (t >> 3) & 0x3f;	/* Bytes already in shsInfo->data */

    /* Handle any leading odd-sized chunks */

    if (t) {
	unsigned char *p = (unsigned char *) ctx->in + t;

	t = 64 - t;
	if (len < t) {
	    memcpy(p, buf, len);
	    return;
	}
	memcpy(p, buf, t);
	byteReverse(ctx->in, 16);
	MD5Transform(ctx->buf, (uint32 *) ctx->in);
	buf += t;
	len -= t;
    }
    /* Process data in 64-byte chunks */

    while (len >= 64) {
	memcpy(ctx->in, buf, 64);
	byteReverse(ctx->in, 16);
	MD5Transform(ctx->buf, (uint32 *) ctx->in);
	buf += 64;
	len -= 64;
    }

    /* Handle any remaining bytes of data. */

    memcpy(ctx->in, buf, len);
}

/*
 * Final wrapup - pad to 64-byte boundary with the bit pattern 
 * 1 0* (64-bit count of bits processed, MSB-first)
 */
void MD5Final(unsigned char digest[16], struct MD5Context *ctx)
{
    unsigned count;
    unsigned char *p;

    /* Compute number of bytes mod 64 */
    count = (ctx->bits[0] >> 3) & 0x3F;

    /* Set the first char of padding to 0x80.  This is safe since there is
       always at least one byte free */
    p = ctx->in + count;
    *p++ = 0x80;

    /* Bytes of padding needed to make 64 bytes */
    count = 64 - 1 - count;

    /* Pad out to 56 mod 64 */
    if (count < 8) {
	/* Two lots of padding:  Pad the first block to 64 bytes */
	memset(p, 0, count);
	byteReverse(ctx->in, 16);
	MD5Transform(ctx->buf, (uint32 *) ctx->in);

	/* Now fill the next block with 56 bytes */
	memset(ctx->in, 0, 56);
    } else {
	/* Pad block to 56 bytes */
	memset(p, 0, count - 8);
    }
    byteReverse(ctx->in, 14);

    /* Append length in bits and transform */
    ((uint32 *) ctx->in)[14] = ctx->bits[0];
    ((uint32 *) ctx->in)[15] = ctx->bits[1];

    MD5Transform(ctx->buf, (uint32 *) ctx->in);
    byteReverse((unsigned char *) ctx->buf, 4);
    memcpy(digest, ctx->buf, 16);
    memset(ctx, 0, sizeof(ctx));	/* In case it's sensitive */
}

#ifndef ASM_MD5

/* The four core functions - F1 is optimized somewhat */

/* #define F1(x, y, z) (x & y | ~x & z) */
#define F1(x, y, z) (z ^ (x & (y ^ z)))
#define F2(x, y, z) F1(z, x, y)
#define F3(x, y, z) (x ^ y ^ z)
#define F4(x, y, z) (y ^ (x | ~z))

/* This is the central step in the MD5 algorithm. */
#define MD5STEP(f, w, x, y, z, data, s) \
	( w += f(x, y, z) + data,  w = w<<s | w>>(32-s),  w += x )

/*
 * The core of the MD5 algorithm, this alters an existing MD5 hash to
 * reflect the addition of 16 longwords of new data.  MD5Update blocks
 * the data and converts bytes into longwords for this routine.
 */
void MD5Transform(uint32 buf[4], uint32 const in[16])
{
    register uint32 a, b, c, d;

    a = buf[0];
    b = buf[1];
    c = buf[2];
    d = buf[3];

    MD5STEP(F1, a, b, c, d,  in[0] + 0xd76aa478U,  7);
    MD5STEP(F1, d, a, b, c,  in[1] + 0xe8c7b756U, 12);
    MD5STEP(F1, c, d, a, b,  in[2] + 0x242070dbU, 17);
    MD5STEP(F1, b, c, d, a,  in[3] + 0xc1bdceeeU, 22);
    MD5STEP(F1, a, b, c, d,  in[4] + 0xf57c0fafU,  7);
    MD5STEP(F1, d, a, b, c,  in[5] + 0x4787c62aU, 12);
    MD5STEP(F1, c, d, a, b,  in[6] + 0xa8304613U, 17);
    MD5STEP(F1, b, c, d, a,  in[7] + 0xfd469501U, 22);
    MD5STEP(F1, a, b, c, d,  in[8] + 0x698098d8U,  7);
    MD5STEP(F1, d, a, b, c,  in[9] + 0x8b44f7afU, 12);
    MD5STEP(F1, c, d, a, b, in[10] + 0xffff5bb1U, 17);
    MD5STEP(F1, b, c, d, a, in[11] + 0x895cd7beU, 22);
    MD5STEP(F1, a, b, c, d, in[12] + 0x6b901122U,  7);
    MD5STEP(F1, d, a, b, c, in[13] + 0xfd987193U, 12);
    MD5STEP(F1, c, d, a, b, in[14] + 0xa679438eU, 17);
    MD5STEP(F1, b, c, d, a, in[15] + 0x49b40821U, 22);

    MD5STEP(F2, a, b, c, d,  in[1] + 0xf61e2562U,  5);
    MD5STEP(F2, d, a, b, c,  in[6] + 0xc040b340U,  9);
    MD5STEP(F2, c, d, a, b, in[11] + 0x265e5a51U, 14);
    MD5STEP(F2, b, c, d, a,  in[0] + 0xe9b6c7aaU, 20);
    MD5STEP(F2, a, b, c, d,  in[5] + 0xd62f105dU,  5);
    MD5STEP(F2, d, a, b, c, in[10] + 0x02441453U,  9);
    MD5STEP(F2, c, d, a, b, in[15] + 0xd8a1e681U, 14);
    MD5STEP(F2, b, c, d, a,  in[4] + 0xe7d3fbc8U, 20);
    MD5STEP(F2, a, b, c, d,  in[9] + 0x21e1cde6U,  5);
    MD5STEP(F2, d, a, b, c, in[14] + 0xc33707d6U,  9);
    MD5STEP(F2, c, d, a, b,  in[3] + 0xf4d50d87U, 14);
    MD5STEP(F2, b, c, d, a,  in[8] + 0x455a14edU, 20);
    MD5STEP(F2, a, b, c, d, in[13] + 0xa9e3e905U,  5);
    MD5STEP(F2, d, a, b, c,  in[2] + 0xfcefa3f8U,  9);
    MD5STEP(F2, c, d, a, b,  in[7] + 0x676f02d9U, 14);
    MD5STEP(F2, b, c, d, a, in[12] + 0x8d2a4c8aU, 20);

    MD5STEP(F3, a, b, c, d,  in[5] + 0xfffa3942U,  4);
    MD5STEP(F3, d, a, b, c,  in[8] + 0x8771f681U, 11);
    MD5STEP(F3, c, d, a, b, in[11] + 0x6d9d6122U, 16);
    MD5STEP(F3, b, c, d, a, in[14] + 0xfde5380cU, 23);
    MD5STEP(F3, a, b, c, d,  in[1] + 0xa4beea44U,  4);
    MD5STEP(F3, d, a, b, c,  in[4] + 0x4bdecfa9U, 11);
    MD5STEP(F3, c, d, a, b,  in[7] + 0xf6bb4b60U, 16);
    MD5STEP(F3, b, c, d, a, in[10] + 0xbebfbc70U, 23);
    MD5STEP(F3, a, b, c, d, in[13] + 0x289b7ec6U,  4);
    MD5STEP(F3, d, a, b, c,  in[0] + 0xeaa127faU, 11);
    MD5STEP(F3, c, d, a, b,  in[3] + 0xd4ef3085U, 16);
    MD5STEP(F3, b, c, d, a,  in[6] + 0x04881d05U, 23);
    MD5STEP(F3, a, b, c, d,  in[9] + 0xd9d4d039U,  4);
    MD5STEP(F3, d, a, b, c, in[12] + 0xe6db99e5U, 11);
    MD5STEP(F3, c, d, a, b, in[15] + 0x1fa27cf8U, 16);
    MD5STEP(F3, b, c, d, a,  in[2] + 0xc4ac5665U, 23);

    MD5STEP(F4, a, b, c, d,  in[0] + 0xf4292244U,  6);
    MD5STEP(F4, d, a, b, c,  in[7] + 0x432aff97U, 10);
    MD5STEP(F4, c, d, a, b, in[14] + 0xab9423a7U, 15);
    MD5STEP(F4, b, c, d, a,  in[5] + 0xfc93a039U, 21);
    MD5STEP(F4, a, b, c, d, in[12] + 0x655b59c3U,  6);
    MD5STEP(F4, d, a, b, c,  in[3] + 0x8f0ccc92U, 10);
    MD5STEP(F4, c, d, a, b, in[10] + 0xffeff47dU, 15);
    MD5STEP(F4, b, c, d, a,  in[1] + 0x85845dd1U, 21);
    MD5STEP(F4, a, b, c, d,  in[8] + 0x6fa87e4fU,  6);
    MD5STEP(F4, d, a, b, c, in[15] + 0xfe2ce6e0U, 10);
    MD5STEP(F4, c, d, a, b,  in[6] + 0xa3014314U, 15);
    MD5STEP(F4, b, c, d, a, in[13] + 0x4e0811a1U, 21);
    MD5STEP(F4, a, b, c, d,  in[4] + 0xf7537e82U,  6);
    MD5STEP(F4, d, a, b, c, in[11] + 0xbd3af235U, 10);
    MD5STEP(F4, c, d, a, b,  in[2] + 0x2ad7d2bbU, 15);
    MD5STEP(F4, b, c, d, a,  in[9] + 0xeb86d391U, 21);

    buf[0] += a;
    buf[1] += b;
    buf[2] += c;
    buf[3] += d;
}

#endif
